prow/cmd/build/controller.go

```diff
@@ -17,46 +17,423 @@ limitations under the License.
+
+	go pjif.Start(stop)
+	go bif.Start(stop)
+
+	if err = controller.Run(2, stop); err != nil {
+		logrus.Fatalf("Error running controller: %v", err)
+	}
+}
+
+type controller struct {
+	pjc prowjobset.Interface
+	bc  buildset.Interface
+
+	pjLister prowjoblisters.ProwJobLister
+	pjSynced cache.InformerSynced
+	bLister  buildlisters.BuildLister
+	bSynced  cache.InformerSynced
+
+	workqueue workqueue.RateLimitingInterface
+
+	recorder record.EventRecorder
+}
+
+func newController(kc kubernetes.Interface, pjc prowjobset.Interface, bc buildset.Interface, pji prowjobinfov1.ProwJobInformer, bi buildinfov1alpha1.BuildInformer) *controller {
+	// Log to events
+	prowjobscheme.AddToScheme(scheme.Scheme)
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(logrus.Infof)
+	eventBroadcaster.StartRecordingToSink(&corev1.EventSinkImpl{Interface: kc.CoreV1().Events("")})
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, untypedcorev1.EventSource{Component: controllerName})
+
+	// Create struct
+	c := &controller{
+		pjc:       pjc,
+		bc:        bc,
+		pjLister:  pji.Lister(),
+		pjSynced:  pji.Informer().HasSynced,
+		bLister:   bi.Lister(),
+		bSynced:   bi.Informer().HasSynced,
+		workqueue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+		recorder:  recorder,
+	}
+
+	logrus.Info("Setting up event handlers")
+
+	// Reconcile whenever a prowjob changes
+	pji.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueKey,
+		UpdateFunc: func(old, new interface{}) {
+			c.enqueueKey(new)
+		},
+		DeleteFunc: c.enqueueKey,
+	})
+
+	// Reconvile whenever a build changes.
+	bi.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.enqueueKey,
+		UpdateFunc: func(old, new interface{}) {
+			c.enqueueKey(new)
+		},
+		DeleteFunc: c.enqueueKey,
+	})
+
+	return c
+}
+
+// Run starts threads workers, returning after receiving a stop signal.
+func (c *controller) Run(threads int, stop <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer c.workqueue.ShutDown()
+
+	logrus.Info("Starting Build controller")
+	logrus.Info("Waiting for informer caches to sync")
+	if ok := cache.WaitForCacheSync(stop, c.pjSynced, c.bSynced); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	logrus.Info("Starting workers")
+	for i := 0; i < threads; i++ {
+		go wait.Until(c.runWorker, time.Second, stop)
+	}
+
+	logrus.Info("Started workers")
+	<-stop
+	logrus.Info("Shutting down workers")
+	return nil
+}
+
+// runWorker dequeues to reconcile, until the queue has closed.
+func (c *controller) runWorker() {
+	for {
+		key, shutdown := c.workqueue.Get()
+		if shutdown {
+			return
+		}
+		defer c.workqueue.Done(key)
+
+		if err := reconcile(c, key.(string)); err != nil {
+			runtime.HandleError(fmt.Errorf("failed to reconcile %s: %v", key, err))
+			continue // Do not forget so we retry later.
```

Comment:
Are you sure this is retried later? When is "later"? I remember in core controllers we had to add back to the queue explicitly.

Reply:
This is how I read the sample-controller docs, unless you call forget it will be requeued (and this is the behavior I've been seeing whenever I had bugs that caused me to fail to create the build):

https://github.com/kubernetes/sample-controller/blob/342d9152cffecfedd97783c86b8074282d019b81/controller.go#L197-L203

```golang
		// We call Done here so the workqueue knows we have finished
		// processing this item. We also must remember to call Forget if we
		// do not want this work item being re-queued. For example, we do
		// not call Forget if a transient error occurs, instead the item is
		// put back on the workqueue and attempted again after a back-off
		// period.
		defer c.workqueue.Done(obj)
```

Reply:
@deads2k can you remind me how the workqueue works? If we don't forget an item and don't requeue on the spot (addratelimited), it is going to be requeued on the next list, is that correct?

Reply:
Sorry for the drive by, but IIUC, calling `Forget()` removes the retry metadata about the item. When `Done()` is called, if the retry metadata is still there, it requeues, otherwise it deletes the item from the queue.